### ISSUE TITLE
unable to build on mac m1 max

### ISSUE THREAD

**Issue by eastonsuo** [zeromq/pyzmq#1907]
*Please use this template for reporting suspected bugs or requests for help.*

# Issue description

when build with no=binary and with bundled 
run
`ZMQ_PREFIX=bundled pip install --no-binary pyzmq pyzmq`

shows error:

`/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/vector:372:5: error: static assertion failed due to requirement 'is_same<zmq::secure_allocator_t<unsigned char>, std::allocator<unsigned char>>::value': [allocator.requirements] states that rebinding an allocator to the same type should result in the original allocato`r

# Environment

* libzmq version (commit hash if unreleased): :pyzmq-25.1.1.tar.gz
* OS: macOS

**Comment by minrk** [comment]
The workaround is to install zeromq prior to installing pyzmq, e.g. via `brew install zeromq`.

This is really a libzmq issue. libzmq 4.3.4 cannot be built without libsodium on current macOS with default parameters, and fails with this error. I'm not sure what compiler flags, if any, can fix it. That's a question for libzmq folks.

The full error, building zeromq 4.3.4 (entirely without pyzmq):

```
export CXXFLAGS=-Wno-deprecated-declarations # avoids warnings turned into errors
./configure --without-libsodium
make V=1
```

gives:

```
Making all in doc
make[1]: Nothing to be done for `all'.
/bin/sh ./libtool  --tag=CXX   --mode=compile g++ -std=gnu++11 -DHAVE_CONFIG_H -I. -I./src           -D_DARWIN_C_SOURCE -D_REENTRANT -D_THREAD_SAFE  -fvisibility=hidden     -Werror -Wall -Wno-long-long -Wno-deprecated-declarations -Wno-atomic-alignment -Wno-tautological-constant-compare -MT src/libzmq_la-ctx.lo -MD -MP -MF src/.deps/libzmq_la-ctx.Tpo -c -o src/libzmq_la-ctx.lo `test -f 'src/ctx.cpp' || echo './'`src/ctx.cpp
libtool: compile:  g++ -std=gnu++11 -DHAVE_CONFIG_H -I. -I./src -D_DARWIN_C_SOURCE -D_REENTRANT -D_THREAD_SAFE -fvisibility=hidden -Werror -Wall -Wno-long-long -Wno-deprecated-declarations -Wno-atomic-alignment -Wno-tautological-constant-compare -MT src/libzmq_la-ctx.lo -MD -MP -MF src/.deps/libzmq_la-ctx.Tpo -c src/ctx.cpp  -fno-common -DPIC -o src/.libs/libzmq_la-ctx.o
libtool: compile:  g++ -std=gnu++11 -DHAVE_CONFIG_H -I. -I./src -D_DARWIN_C_SOURCE -D_REENTRANT -D_THREAD_SAFE -fvisibility=hidden -Werror -Wall -Wno-long-long -Wno-deprecated-declarations -Wno-atomic-alignment -Wno-tautological-constant-compare -MT src/libzmq_la-ctx.lo -MD -MP -MF src/.deps/libzmq_la-ctx.Tpo -c src/ctx.cpp -o src/libzmq_la-ctx.o >/dev/null 2>&1
mv -f src/.deps/libzmq_la-ctx.Tpo src/.deps/libzmq_la-ctx.Plo
/bin/sh ./libtool  --tag=CXX   --mode=compile g++ -std=gnu++11 -DHAVE_CONFIG_H -I. -I./src           -D_DARWIN_C_SOURCE -D_REENTRANT -D_THREAD_SAFE  -fvisibility=hidden     -Werror -Wall -Wno-long-long -Wno-deprecated-declarations -Wno-atomic-alignment -Wno-tautological-constant-compare -MT src/libzmq_la-curve_client.lo -MD -MP -MF src/.deps/libzmq_la-curve_client.Tpo -c -o src/libzmq_la-curve_client.lo `test -f 'src/curve_client.cpp' || echo './'`src/curve_client.cpp
libtool: compile:  g++ -std=gnu++11 -DHAVE_CONFIG_H -I. -I./src -D_DARWIN_C_SOURCE -D_REENTRANT -D_THREAD_SAFE -fvisibility=hidden -Werror -Wall -Wno-long-long -Wno-deprecated-declarations -Wno-atomic-alignment -Wno-tautological-constant-compare -MT src/libzmq_la-curve_client.lo -MD -MP -MF src/.deps/libzmq_la-curve_client.Tpo -c src/curve_client.cpp  -fno-common -DPIC -o src/.libs/libzmq_la-curve_client.o
In file included from src/curve_client.cpp:35:
In file included from ./src/msg.hpp:40:
In file included from ./src/metadata.hpp:33:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/map:2356:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/functional:515:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/__functional/boyer_moore_searcher.h:26:
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/vector:372:5: error: static assertion failed due to requirement 'is_same<zmq::secure_allocator_t<unsigned char>, std::allocator<unsigned char>>::value': [allocator.requirements] states that rebinding an allocator to the same type should result in the original allocator
    static_assert(is_same<allocator_type, __rebind_alloc<__alloc_traits, value_type> >::value,
    ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./src/curve_client_tools.hpp:64:60: note: in instantiation of template class 'std::vector<unsigned char, zmq::secure_allocator_t<unsigned char>>' requested here
        std::vector<uint8_t, secure_allocator_t<uint8_t> > hello_plaintext (
                                                           ^
./src/curve_client_tools.hpp:110:60: error: no matching constructor for initialization of 'std::vector<uint8_t, secure_allocator_t<uint8_t>>' (aka 'vector<unsigned char, secure_allocator_t<unsigned char>>')
        std::vector<uint8_t, secure_allocator_t<uint8_t> > welcome_plaintext (
                                                           ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/vector:348:28: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'int' to 'const std::vector<unsigned char, zmq::secure_allocator_t<unsigned char>>' for 1st argument
class _LIBCPP_TEMPLATE_VIS vector
                           ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/vector:348:28: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'int' to 'std::vector<unsigned char, zmq::secure_allocator_t<unsigned char>>' for 1st argument
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/vector:348:28: note: candidate constructor (the implicit default constructor) not viable: requires 0 arguments, but 1 was provided
In file included from src/curve_client.cpp:38:
In file included from ./src/curve_client.hpp:37:
./src/curve_client_tools.hpp:121:53: error: type 'std::vector<uint8_t, secure_allocator_t<uint8_t>>' (aka 'vector<unsigned char, secure_allocator_t<unsigned char>>') does not provide a subscript operator
        int rc = crypto_box_open (&welcome_plaintext[0], welcome_box,
                                   ~~~~~~~~~~~~~~~~~^~
./src/curve_client_tools.hpp:129:47: error: type 'std::vector<uint8_t, secure_allocator_t<uint8_t>>' (aka 'vector<unsigned char, secure_allocator_t<unsigned char>>') does not provide a subscript operator
        memcpy (cn_server_, &welcome_plaintext[crypto_box_ZEROBYTES], 32);
                             ~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~
./src/curve_client_tools.hpp:130:47: error: type 'std::vector<uint8_t, secure_allocator_t<uint8_t>>' (aka 'vector<unsigned char, secure_allocator_t<unsigned char>>') does not provide a subscript operator
        memcpy (cn_cookie_, &welcome_plaintext[crypto_box_ZEROBYTES + 32],
                             ~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~
./src/curve_client_tools.hpp:154:60: error: no matching constructor for initialization of 'std::vector<uint8_t, secure_allocator_t<uint8_t>>' (aka 'vector<unsigned char, secure_allocator_t<unsigned char>>')
        std::vector<uint8_t, secure_allocator_t<uint8_t> > vouch_plaintext (
                                                           ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/vector:348:28: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'int' to 'const std::vector<unsigned char, zmq::secure_allocator_t<unsigned char>>' for 1st argument
class _LIBCPP_TEMPLATE_VIS vector
                           ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/vector:348:28: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'int' to 'std::vector<unsigned char, zmq::secure_allocator_t<unsigned char>>' for 1st argument
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/vector:348:28: note: candidate constructor (the implicit default constructor) not viable: requires 0 arguments, but 1 was provided
In file included from src/curve_client.cpp:38:
In file included from ./src/curve_client.hpp:37:
./src/curve_client_tools.hpp:159:36: error: no member named 'begin' in 'std::vector<unsigned char, zmq::secure_allocator_t<unsigned char>>'
        std::fill (vouch_plaintext.begin (),
                   ~~~~~~~~~~~~~~~ ^
./src/curve_client_tools.hpp:160:36: error: no member named 'begin' in 'std::vector<unsigned char, zmq::secure_allocator_t<unsigned char>>'
                   vouch_plaintext.begin () + crypto_box_ZEROBYTES, 0);
                   ~~~~~~~~~~~~~~~ ^
./src/curve_client_tools.hpp:161:33: error: type 'std::vector<uint8_t, secure_allocator_t<uint8_t>>' (aka 'vector<unsigned char, secure_allocator_t<unsigned char>>') does not provide a subscript operator
        memcpy (&vouch_plaintext[crypto_box_ZEROBYTES], cn_public_, 32);
                 ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~
./src/curve_client_tools.hpp:162:33: error: type 'std::vector<uint8_t, secure_allocator_t<uint8_t>>' (aka 'vector<unsigned char, secure_allocator_t<unsigned char>>') does not provide a subscript operator
        memcpy (&vouch_plaintext[crypto_box_ZEROBYTES + 32], server_key_, 32);
                 ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~
./src/curve_client_tools.hpp:169:50: error: type 'std::vector<uint8_t, secure_allocator_t<uint8_t>>' (aka 'vector<unsigned char, secure_allocator_t<unsigned char>>') does not provide a subscript operator
          crypto_box (vouch_box, &vouch_plaintext[0], vouch_plaintext.size (),
                                  ~~~~~~~~~~~~~~~^~
./src/curve_client_tools.hpp:169:71: error: no member named 'size' in 'std::vector<unsigned char, zmq::secure_allocator_t<unsigned char>>'
          crypto_box (vouch_box, &vouch_plaintext[0], vouch_plaintext.size (),
                                                      ~~~~~~~~~~~~~~~ ^
./src/curve_client_tools.hpp:177:60: error: no matching constructor for initialization of 'std::vector<uint8_t, secure_allocator_t<uint8_t>>' (aka 'vector<unsigned char, secure_allocator_t<unsigned char>>')
        std::vector<uint8_t, secure_allocator_t<uint8_t> > initiate_plaintext (
                                                           ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/vector:348:28: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'size_t' (aka 'unsigned long') to 'const std::vector<unsigned char, zmq::secure_allocator_t<unsigned char>>' for 1st argument
class _LIBCPP_TEMPLATE_VIS vector
                           ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/vector:348:28: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'size_t' (aka 'unsigned long') to 'std::vector<unsigned char, zmq::secure_allocator_t<unsigned char>>' for 1st argument
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/vector:348:28: note: candidate constructor (the implicit default constructor) not viable: requires 0 arguments, but 1 was provided
In file included from src/curve_client.cpp:38:
In file included from ./src/curve_client.hpp:37:
./src/curve_client_tools.hpp:181:39: error: no member named 'begin' in 'std::vector<unsigned char, zmq::secure_allocator_t<unsigned char>>'
        std::fill (initiate_plaintext.begin (),
                   ~~~~~~~~~~~~~~~~~~ ^
./src/curve_client_tools.hpp:182:39: error: no member named 'begin' in 'std::vector<unsigned char, zmq::secure_allocator_t<unsigned char>>'
                   initiate_plaintext.begin () + crypto_box_ZEROBYTES, 0);
                   ~~~~~~~~~~~~~~~~~~ ^
./src/curve_client_tools.hpp:183:36: error: type 'std::vector<uint8_t, secure_allocator_t<uint8_t>>' (aka 'vector<unsigned char, secure_allocator_t<unsigned char>>') does not provide a subscript operator
        memcpy (&initiate_plaintext[crypto_box_ZEROBYTES], public_key_, 32);
                 ~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~
./src/curve_client_tools.hpp:184:36: error: type 'std::vector<uint8_t, secure_allocator_t<uint8_t>>' (aka 'vector<unsigned char, secure_allocator_t<unsigned char>>') does not provide a subscript operator
        memcpy (&initiate_plaintext[crypto_box_ZEROBYTES + 32], vouch_nonce + 8,
                 ~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~
./src/curve_client_tools.hpp:186:36: error: type 'std::vector<uint8_t, secure_allocator_t<uint8_t>>' (aka 'vector<unsigned char, secure_allocator_t<unsigned char>>') does not provide a subscript operator
        memcpy (&initiate_plaintext[crypto_box_ZEROBYTES + 48],
                 ~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~
./src/curve_client_tools.hpp:189:40: error: type 'std::vector<uint8_t, secure_allocator_t<uint8_t>>' (aka 'vector<unsigned char, secure_allocator_t<unsigned char>>') does not provide a subscript operator
            memcpy (&initiate_plaintext[crypto_box_ZEROBYTES + 48 + 80],
                     ~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
make[1]: *** [src/libzmq_la-curve_client.lo] Error 1
make: *** [all-recursive] Error 1
```


I think libzmq 4.3.5 won't have this issue simply because the affected feature has been removed. This, in turn, is part of what's blocking pyzmq's upgrade to 4.3.5, since it's a major breaking change in pyzmq distributions to lose CURVE support. So I have to figure out how to bundle libsodium and replace the abandoned Windows builds of libzmq, which don't exist for 4.3.5 before pyzmq can update the bundled libzmq. So I think this will be closed by #1903.

**Comment by eastonsuo** [comment]
if I use `brew install zeromq` first ,will `ZMQ_PREFIX=bundled` works if I want to build an extension?

**Comment by minrk** [comment]
Ah, no, it should link to the installed libzmq (omit `ZMQ_PREFIX=bundled`, or it will ignore the installed libzmq). But that's usually ought to be a good thing. If you want to repack it for redistribution, use `delocate`, which will bring the linked libzmq into the wheel. This will ~always produce better results than the bundled libzmq, which is used only as a fallback when the a desirable 'real' build of libzmq is unavailable.

**Comment by tacaswell** [comment]
I am also seeing this error on fedora 38

```
✘ 11:27:28 [yennefer] $ g++ -std=c++17 -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -DZMQ_HAVE_CURVE=1 -DZMQ_USE_TWEETNACL=1 -DZMQ_USE_EPOLL=1 -DZMQ_IOTHREADS_USE_EPOLL=1 -DZMQ_POLL_BA
SED_ON_POLL=1 -Ibundled/zeromq/include -Ibundled -I/home/tcaswell/.virtualenvs/bleeding/include -I/home/tcaswell/.pybuild/bleeding/include/python3.11 -c bundled/zeromq/src/curve_client.cpp -
o build/temp.linux-x86_64-cpython-311/bundled/zeromq/src/curve_client.o
In file included from /usr/include/c++/13/ext/alloc_traits.h:34,
                 from /usr/include/c++/13/bits/stl_tree.h:67,
                 from /usr/include/c++/13/map:62,
                 from bundled/zeromq/src/metadata.hpp:33,
                 from bundled/zeromq/src/msg.hpp:40,
                 from bundled/zeromq/src/curve_client.cpp:35:
/usr/include/c++/13/bits/alloc_traits.h: In instantiation of ‘struct std::__allocator_traits_base::__rebind<zmq::secure_allocator_t<unsigned char>, unsigned char, void>’:
/usr/include/c++/13/bits/alloc_traits.h:94:11:   required by substitution of ‘template<class _Alloc, class _Up> using std::__alloc_rebind = typename std::__allocator_traits_base::__rebind<_Alloc, _Up>::type [with _Alloc = zmq::secure_allocator_t<unsigned char>; _Up = unsigned char]’
/usr/include/c++/13/bits/alloc_traits.h:228:8:   required by substitution of ‘template<class _Alloc> template<class _Tp> using std::allocator_traits< <template-parameter-1-1> >::rebind_alloc = std::__alloc_rebind<_Alloc, _Tp> [with _Tp = unsigned char; _Alloc = zmq::secure_allocator_t<unsigned char>]’
/usr/include/c++/13/ext/alloc_traits.h:126:65:   required from ‘struct __gnu_cxx::__alloc_traits<zmq::secure_allocator_t<unsigned char>, unsigned char>::rebind<unsigned char>’
/usr/include/c++/13/bits/stl_vector.h:88:21:   required from ‘struct std::_Vector_base<unsigned char, zmq::secure_allocator_t<unsigned char> >’
/usr/include/c++/13/bits/stl_vector.h:425:11:   required from ‘class std::vector<unsigned char, zmq::secure_allocator_t<unsigned char> >’
bundled/zeromq/src/curve_client_tools.hpp:64:76:   required from here
/usr/include/c++/13/bits/alloc_traits.h:70:31: error: static assertion failed: allocator_traits<A>::rebind_alloc<A::value_type> must be A
   70 |                         _Tp>::value,
      |                               ^~~~~
/usr/include/c++/13/bits/alloc_traits.h:70:31: note: ‘std::integral_constant<bool, false>::value’ evaluates to false

```

Confirm that installing zeromq-devel avoids the problem.


```
$ gcc --version
gcc (GCC) 13.2.1 20231011 (Red Hat 13.2.1-4)
Copyright (C) 2023 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


```

**Comment by minrk** [comment]
I dread it a bit, but I feel like this and the removal of tweetnacl from libzmq is going to force me to rewrite bundling to include 'proper' installation of libzmq and libsodium and building via something like scikit-build or meson. I think that will be better in the long run, but figuring out the linking on Windows is going to take a _lot_ of trial and error, and certainly reintroduce lots of build failures fixed over the years on less common platforms.

If anyone with C++ knowledge knows what's changed with recent compilers to introduce these failures and what flags and/or patches will keep us limping along while I figure that out, it would be much appreciated and take a lot of pressure off.

**Comment by minrk** [comment]
upstream fix appears to be https://github.com/zeromq/libzmq/pull/4480 I'll see if I can apply a patch to the bundled sources

**Comment by minrk** [comment]
If anyone affected can test with the sdists from [this run](https://github.com/zeromq/pyzmq/actions/runs/7057525142) and confirm that it fixes the issue, I can merge #1913 and make a 25.1.2 with it.

**Comment by minrk** [comment]
this should be fixed with last month's 25.1.2
